contracts/CreditVoucher.sol

```diff
@@ -0,0 +1,178 @@
+import {SignatureChecker} from "@openzeppelin/contracts/utils/cryptography/SignatureChecker.sol";
+import {Pausable} from "@openzeppelin/contracts/utils/Pausable.sol";
+
+import {ICreditVoucher} from "./interfaces/ICreditVoucher.sol";
+
+contract CreditVoucher is ICreditVoucher, ERC721Enumerable, Pausable, EIP712, Ownable {
+    using SafeERC20 for IERC20;
+    using SignatureChecker for address;
+
+    bytes32 public constant VERSION = keccak256("CREDIT_VOUCHER_V1");
+    // keccak256("Mint(address to,uint256 creditValue,uint256 nonce,uint256 deadline)")
+    bytes32 private constant MINT_TYPEHASH = 0xd10efb11fc8417faf3162be9fb4bccfb2fe444b18748b7f8b5693464fbaa56bb;
+
+    address public immutable USDC;
+
+    address public operator;
+    uint256 public validityDuration;
+    string public baseTokenURI;
+
+    mapping(uint256 => Voucher) public vouchers;
+
+    // the nonces mapping is given for replay protection
+    mapping(address => uint256) public sigNonces;
+
+    constructor(
+        address _operator,
+        address _usdc,
+        string memory _uri
+    ) ERC721("CreditVoucher", "CV") EIP712("TRVLCreditVoucher", "1") Ownable(_msgSender()) {
+        if (_operator == address(0) || _usdc == address(0)) {
+            revert ZeroAddress();
+        }
+        operator = _operator;
+        USDC = _usdc;
+        baseTokenURI = _uri;
+
+        // Token transfers are disabled by default, except for minting/burning
+        _pause();
+    }
+
+    function _update(address to, uint256 tokenId, address auth) internal override(ERC721Enumerable) returns (address) {
+        if (paused()) {
+            if (_ownerOf(tokenId) != address(0) || to != address(0)) {
+                revert TransferWhilePaused();
+            }
+        }
+        return super._update(to, tokenId, auth);
+    }
+
+    function _baseURI() internal view override returns (string memory) {
+        return baseTokenURI;
+    }
+
+    /**
+     * @notice Set operator address
+     * @dev    Caller must be CONTRACT OWNER
+     * @param _addr The new operator address
+     */
+    function setOperator(address _addr) external onlyOwner {
+        operator = _addr;
+        emit NewOperator(_addr);
+    }
+
+    /**
+     * @notice Set base token URI
+     * @dev    Caller must be CONTRACT OWNER or OPERATOR
+     * @param _uri Th new token base URI
+     */
+    function setBaseTokenURI(string calldata _uri) external {
+        address msgSender = _msgSender();
+        if (msgSender != owner() && msgSender != operator) {
+            revert Unauthorized();
+        }
+        baseTokenURI = _uri;
+        emit NewBaseTokenURI(_uri);
+    }
+
+    /**
+        @notice Set validity duration
+        @dev Caller must be CONTRACT OWNER
+        @param _duration the new validity duration
+    */
+    function setValidityDuration(uint256 _duration) external onlyOwner {
+        validityDuration = _duration;
```

Comment:
Instead of having a global validity duration, let's move this to the Voucher struct, to allow each voucher to have its own validity.